# for paper, lets start with 1706 NMD genes expressed in LCLs
library(dplyr)
library(xlsx)
library(VennDiagram)
library(extrafont)
loadfonts()
source('R/pubTheme.R')

# expressed 1706 genes
load('data/mend_genes_isoform_analysis_input.RData')
gtex.ebv <- gtex.ebv[which(gtex.ebv$mend.gene == "Y"),]
length(unique(gtex.ebv$gene_symbol)) # 1706 gene symbols
ids <- ebv.expr[which(ebv.expr$gene_symbol %in% gtex.ebv$gene_symbol),'gene_id'] # 1706 gene ids

# generated by /mnt/toil_20k/data/projects/cdls_analysis/get_meanCov_meanTPM.R
ebv.tpm <- readRDS('data/EBV_meanTPMandCoverage.RDS')
ebv.tpm <- ebv.tpm[which(ebv.tpm$gene_id %in% ids),]
ebv.tpm$expressed_transcript <- ifelse(ebv.tpm$meanCov > 10 & ebv.tpm$meanTPM > 1, "Yes", "No")
plyr::count(ebv.tpm$expressed_transcript) # 6628/17894 are actually expressed = 37%

# use the 17894 LCL transcripts to correlate between Brain and LCLs
# load brain TPM data
brain.tpm <- readRDS('data/Brain_meanTPMandCoverage.RDS')
brain.tpm <- brain.tpm[which(brain.tpm$gene_id %in% ids),]
brain.tpm$expressed_transcript <- ifelse(brain.tpm$meanCov > 10 & brain.tpm$meanTPM > 1, "Yes", "No")
plyr::count(brain.tpm$expressed_transcript) # 5278/17894 are expressed = 29%

# intersect
# overlap between all expressed transcripts?
ebv.tns <- unique(ebv.tpm[which(ebv.tpm$expressed_transcript == "Yes"),"transcript_id"])
brain.tns <- unique(brain.tpm[which(brain.tpm$expressed_transcript == "Yes"),"transcript_id"])
length(intersect(ebv.tns, brain.tns)) # 4182/6628 = 63%

# annotate max expressed transcript for brain
# overlap between max expressed transcript?
brain.tpm <- brain.tpm %>% group_by(gene_id) %>%
  mutate(Brain_max = ifelse(meanTPM == max(meanTPM), 1, 0)) %>%
  as.data.frame()
brain.max <- brain.tpm[which(brain.tpm$Brain_max == 1),'transcript_id']
ebv.tpm <- ebv.tpm %>% group_by(gene_id) %>%
  mutate(EBV_max = ifelse(meanTPM == max(meanTPM), 1, 0)) %>%
  as.data.frame()
ebv.max <- ebv.tpm[which(ebv.tpm$EBV_max == 1),'transcript_id']
length(intersect(brain.max, ebv.max)) # 1114/1706 = 65%

# supplementary table 2 (Table S2)
colnames(brain.tpm) <- c('gene_id','transcript_id','Brain_meanCov','Brain_meanTPM','gene_symbol','Brain_expressed','Brain_max_expressed')
ebv.tpm$body_site <- NULL
colnames(ebv.tpm) <- c('gene_id','transcript_id','EBV_meanCov','EBV_meanTPM','gene_symbol','EBV_expressed','EBV_max_expressed')
total <- merge(ebv.tpm, brain.tpm, by = c('gene_id','gene_symbol','transcript_id'))
write.xlsx(x = total, file = 'manuscript/Table_S2.xlsx', sheetName = "Supp_Table_2", row.names = F)

# scatter plot between EBV and Brain transcripts  (Fig 1d)
p.cor <- cor.test(log2(total$EBV_meanTPM+1), log2(total$Brain_meanTPM+1), method = "pearson")
p <- ggplot(total, aes(log2(EBV_meanTPM+1), log2(Brain_meanTPM+1))) +
  geom_point(pch = 21, size = 1) + geom_smooth(method=lm, size = 0.5, color = 'darkgray') +
  geom_hline(yintercept = log2(1+1), linetype = "dashed", color = "red") +
  geom_vline(xintercept = log2(1+1), linetype = "dashed", color = "red") + theme_Publication4() +
  xlab('EBV log2(TPM)') + ylab('Brain log2(TPM)') +
  annotate(geom = "text", x = 5, y = 10, label = paste0("Pearson: 0.74\n Pval:<2.2e-16"), color = "red", size = 3)
p <- p + ggtitle('NMD Genes: Correlation of Transcript Expression\n(nGenes = 1706, nTranscripts = 17894)')
ggsave(filename = 'paper/mendelian_disorders_genes/nmd_genes_isoform_expression_LCLvsBrain_Fig1d.pdf', device = "pdf", plot = p, width = 6, height = 6)

# Venn diagrams (Fig 1e, 1f)
# Isoform expression overlap (Fig 1e)
g <- draw.pairwise.venn(area1 = length(ebv.tns), area2 = length(brain.tns),
                        cross.area = length(intersect(brain.tns, ebv.tns)),
                        scaled = FALSE, category = c("LCL\n(6628)", "Brain\n(5278)"),
                        fill = c("pink1","skyblue"), alpha = c(0.7,0.7),
                        cat.pos = c(-20,20), cat.dist = c(0.06, 0.06),
                        fontfamily = rep("Arial", 3),
                        cex = rep(1, 3),
                        cat.cex = rep(1, 2),
                        cat.fontfamily = rep("Arial", 2))
tg <- textGrob("Isoform expression overlap", gp = gpar(fontsize = 12, fontfamily = "Arial"))
sg <- textGrob("(FPKM >1 & mean coverage >10x)", gp = gpar(fontsize = 10, fontface = 3L, fontfamily = "Arial"))
margin <- unit(0.5, "line")
grid.newpage()
g <- grid.arrange(tg, sg, gTree(children=g),
             heights = unit.c(grobHeight(tg) + 1.2*margin,
                              grobHeight(sg) + margin,
                              unit(1,"null")))
ggsave("manuscript/venn_isoform_expr_overlap_Fig1d.pdf", plot = g, device = "pdf", width = 5, height = 4)


# Max expression isoform overlap (Fig 1f)
g <- draw.pairwise.venn(area1 = length(ebv.max), area2 = length(brain.max),
                        cross.area = length(intersect(brain.max, ebv.max)),
                        scaled = FALSE, category = c("LCL\n(1706)", "Brain\n(1706)"),
                        fill = c("pink1","skyblue"), alpha = c(0.7,0.7),
                        cat.pos = c(-20,20), cat.dist = c(0.06, 0.06),
                        fontfamily = rep("Arial", 3),
                        cex = rep(1, 3),
                        cat.cex = rep(1, 2),
                        cat.fontfamily = rep("Arial", 2))
tg <- textGrob("Max expressed isoform overlap", gp=gpar(fontsize = 12, fontfamily = "Arial"))
margin <- unit(0.5, "line")
grid.newpage()
g <- grid.arrange(tg, gTree(children=g),
             heights = unit.c(grobHeight(tg) + margin,
                              unit(1,"null")))
ggsave("manuscript/venn_max_expr_isoform_Fig1e.pdf", plot = g, device = "pdf", width = 5, height = 4)
